### PR TITLE
fixing #47 filename finder in  download get_dependency.py

### PR DESCRIPTION
**Pull request by amit2rockon**
solving #47

### CODE REVIEW

File: get_dependencies.py [hunk 1]
```diff
@@ -5,7 +5,8 @@
 import requests
 
 
-def download(url, filename):
+def download(url):
+    filename = url[url.rfind("/") + 1:url.find("?")]
```
Comment (marco-c) [inline get_dependencies.py]: This is a bit brittle and could break if we start using other URLs.
We should use the `Content-Disposition` header of the response and, if it doesn't exist, use the basename of the URL.

File: get_dependencies.py [hunk 2]
```diff
@@ -26,17 +27,14 @@ def download(url, filename):
...
-    name = 'mac.tar.xz'
 elif sys.platform.startswith('win32'):
     url = 'https://www.dropbox.com/s/xskj9rpn2fjkra8/win32.tar.xz?dl=1'
-    name = 'win32.tar.xz'
 
 print('[*] Downloading support files...')
-download(url, name)
-
+download(url)
+name = url[url.rfind("/") + 1:url.find("?")]
```
Comment (marco-c) [inline get_dependencies.py]: Make `download` return the name, so we avoid duplicating the code here.

File: get_dependencies.py (hunk 1 above)
Comment (amit2rockon) [reply get_dependencies.py]: what about the case when the url will not contain the file name?

File: get_dependencies.py (hunk 2 above)
Comment (amit2rockon) [reply get_dependencies.py]: fname = download(url)
 tarfile.open(fname, 'r:xz')
....
os.remove(fname)
may i use this way if download return the filename?

File: get_dependencies.py (hunk 1 above)
Comment (marco-c) [reply get_dependencies.py]: if there's no `Content-Disposition` header and the basename is an empty string, throw an exception (`Couldn't get file name for this URL`, or something like this).

File: get_dependencies.py (hunk 2 above)
Comment (marco-c) [reply get_dependencies.py]: Why do you want to change the name of the variable from `name` to `fname`?

File: get_dependencies.py (hunk 1 above)
Comment (amit2rockon) [reply get_dependencies.py]: okay

File: get_dependencies.py (hunk 2 above)
Comment (amit2rockon) [reply get_dependencies.py]: variable will be name only , then is this permissible  assigning download(url) to name?

Comment (amit2rockon) [reply get_dependencies.py]: apologies me , but just asking that download(url) will download file though it will return filename yet assigning to a variable named 'name' may cause to misunderstand the variable (for the new ones).

Comment (marco-c) [reply get_dependencies.py]: I still don't understand the question 😄 
Let's leave the name of the variable as `name`, we can consider changing it in the future in a separate PR.

Comment (amit2rockon) [reply get_dependencies.py]: okay

**Comment by amit2rockon** [comment]
@marco-c  review it.

File: get_dependencies.py [hunk 3]
```diff
@@ -5,9 +5,11 @@
 import requests
 
 
-def download(url, filename):
+def download(url):
+    response = requests.get(url, stream=True)
+    filename = response.headers.get('content-disposition')
+    filename = filename[filename.find("\"") + 1: filename.rfind("\"")]
```
Comment (marco-c) [inline get_dependencies.py]: This is not implementing the fallback when the content-disposition header doesn't exist.
I suggest you use the `rfc6266` library to do this.

Comment (konemshad) [reply get_dependencies.py]: @marco-c   i think the pr of mine which you closed recently implements your requirement, doesn't it?

Comment (marco-c) [reply get_dependencies.py]: I don't think so, in your PR you're printing `File Not Found` when the content-disposition header doesn't exist and the `filename` variable is not defined in that case.

Comment (konemshad) [reply get_dependencies.py]: OK. what should we store the value in the variable "filename" if the content-disposition header does not exist ? we can just add an else clause?

Comment (marco-c) [reply get_dependencies.py]: Discussion at https://github.com/marco-c/autowebcompat/pull/48#discussion_r167229099.

**Comment by ManasviKundalia** [comment]
[@marco-c is this okay? ](https://github.com/ManasviKundalia/autowebcompat/blob/c48b9d2693ea6e0ff175792837a111a13d8af701/get_dependencies.py)

**Comment by amit2rockon** [comment]
okay

**Comment by marco-c** [comment]
> @marco-c is this okay? 

There are some problems with your solution:
1) It's using filename_unsafe, which is, as the name says, unsafe 😄 
2) It's not implementing the fallback of using the basename of the file from the URL when the header doesn't exist;
3) I would just use parse_requests_response directly

**Comment by ManasviKundalia** [comment]
@marco-c I have made the changes [here](https://github.com/ManasviKundalia/autowebcompat/blob/ManasviKundalia-patch-1/get_dependencies.py) 
Do you have any examples/ corner cases on which I could test the code?

**Comment by amit2rockon** [comment]
@marco-c  can it be generalized for every link even if the link is containing the downloadable file?
like this link.    
https://drive.google.com/uc?export=download&confirm=oIyQ&id=17sqc0GQlUiFHej-OZPasJS5ionZkoCK1
though in the case of google drive google adds BOM (Byte Order Marks) to the beginning of some files.
 and this link contains linux.zip (old file) even I tried @ManasviKundalia last committed code and it is actually getting failed at this point.
so if the file is going to be only on dropbox then its okay but what if such link is present?

**Comment by amit2rockon** [comment]
@marco-c  where to upload the downloaded extra screenshots? 
and what if other's downloaded screenshots will have the same name but different files?

**Comment by marco-c** [comment]
> @marco-c can it be generalized for every link even if the link is containing the downloadable file?

I don't understand this question, sorry. Could you rephrase?
Maybe you mean what should we do when the `Content-Disposition` header is missing and there's no basename in the URL? I answered to this question in https://github.com/marco-c/autowebcompat/pull/48#discussion_r167235321.

**Comment by marco-c** [comment]
> @marco-c where to upload the downloaded extra screenshots?
and what if other's downloaded screenshots will have the same name but different files?

Which extra screenshots?

**Comment by amit2rockon** [comment]
extra screenshots means ,the screenshots which are not present in current data.zip

**Comment by marco-c** [comment]
> @marco-c I have made the changes here
Do you have any examples/ corner cases on which I could test the code?

With `parse_requests_response`, you don't have to parse the content-disposition header yourself. Also, you can drop the except case, `parse_requests_response` automatically implements the basename fallback.
I didn't notice that `filename_sanitized` requires an extension. Please use `filename_unsafe` then and sanitize it manually like `filename_sanitized` does (at https://github.com/g2p/rfc6266/blob/master/rfc6266.py#L121) except for the extension part.

**Comment by marco-c** [comment]
> extra screenshots means ,the screenshots which are not present in current data.zip

Whenever we have new screenshots, we will upload a new data.zip file. Anyway, this seems unrelated to the current PR.

**Comment by amit2rockon** [comment]
okay .

**Comment by amit2rockon** [comment]
How to handle the case when the link is broken but contains baseurl ?
`parse_requests_response `  is assigning the  filename with  baseurl of the broken link also.

**Comment by ManasviKundalia** [comment]
@marco-c I did the changes like you suggested replacing filename_sanitized with filename_unsafe and later sanitizing without using the extension. Please [review](https://github.com/ManasviKundalia/autowebcompat/blob/ManasviKundalia-patch-1/get_dependencies.py).

**Comment by amit2rockon** [comment]
@marco-c   see these [changes  ](https://github.com/amit2rockon/autowebcompat/blob/master/get_dependencies.py)
it also solves the condition when the link is broken while it contain the baseurl.

File: get_dependencies.py [hunk 4]
```diff
@@ -3,15 +3,25 @@
 import tarfile
 from zipfile import ZipFile
 import requests
+from rfc6266 import parse_requests_response
 
+def download(url):
+    response = requests.get(url, stream=True)
+    total = response.headers.get('content-length')
+    filename = parse_requests_response(response).filename_unsafe
```
Comment (marco-c) [inline get_dependencies.py]: As I said in https://github.com/marco-c/autowebcompat/pull/48#issuecomment-365648565, this needs to be sanitized.

**Comment by marco-c** [comment]
> How to handle the case when the link is broken but contains baseurl ?

Ignore this case, this PR is only about making the `download` function guess the name.

**Comment by marco-c** [comment]
> @marco-c I did the changes like you suggested replacing filename_sanitized with filename_unsafe and later sanitizing without using the extension. Please review.

It looks good (modulo some minor issues), thanks. @ManasviKundalia anyway, since this issue is already being worked on by @amit2rockon, I suggest you pick up some other work item from the issues list.
Unfortunately only one person at a time can work on any given issue

File: get_dependencies.py [hunk 5]
```diff
@@ -3,15 +3,33 @@
...
+from rfc6266 import parse_requests_response
+import posixpath
 
 
-def download(url, filename):
+def download(url):
+    response = requests.get(url, stream=True)
+    filename = parse_requests_response(response).filename_unsafe
+
+    default_filename = 'file'
```
Comment (marco-c) [inline get_dependencies.py]: Move the sanitizing into a separate `sanitize` function.

File: get_dependencies.py [hunk 6]
```diff
@@ -3,15 +3,33 @@
...
+    if filename is None:
+        filename = default_filename
+    filename = posixpath.basename(filename)
+    filename = os.path.basename(filename)
+    filename = filename.lstrip('.')
+    if not filename:
+        filename = default_filename
+
+    if filename.rfind('.') == -1:
```
Comment (marco-c) [inline get_dependencies.py]: Remove this, we don't care if the file doesn't have an extension.

File: get_dependencies.py [hunk 7]
```diff
@@ -3,15 +3,33 @@
...
+        print('Couldn\'t get filename for this url')
+        return
+
     with open(filename, 'wb') as f:
-        response = requests.get(url, stream=True)
         total = response.headers.get('content-length')
 
         if total is None:
             f.write(response.content)
+            return
```
Comment (marco-c) [inline get_dependencies.py]: Why are you returning here?

File: get_dependencies.py [hunk 8]
```diff
@@ -22,32 +40,36 @@ def download(url, filename):
...
-with tarfile.open(name, 'r:xz') as f:
-    f.extractall('.')
-os.remove(name)
+try:
+    name = download(url)
+    print('[*] Extracting files...')
+    with tarfile.open(name, 'r:xz') as f:
+        f.extractall('.')
+    os.remove(name)
+except ValueError as e:
```
Comment (marco-c) [inline get_dependencies.py]: We don't want to catch exceptions here. If there's an exception, we want the script to fail.

File: get_dependencies.py [hunk 9]
```diff
@@ -22,32 +40,36 @@ def download(url, filename):
...
 
-print('[*] Extracting data.zip...')
-with ZipFile('data.zip', 'r') as z:
-    z.extractall()
+try:
+    with ZipFile('data.zip', 'r') as z:
+        print('[*] Extracting data.zip...')
+        z.extractall()
+    os.remove('data.zip')
+except ValueError as e:
```
Comment (marco-c) [inline get_dependencies.py]: Same here.

File: requirements.txt [hunk 10]
```diff
@@ -4,3 +4,4 @@ keras==2.1.1
 tensorflow==1.3.0
 h5py==2.7.1
 requests==2.18.4
+rfc6266==0.0.4
```
Comment (marco-c) [inline requirements.txt]: This should go in `test-requirements.txt`, as it's only needed for development.

File: get_dependencies.py [hunk 11]
```diff
@@ -3,15 +3,33 @@
...
+import posixpath
 
 
-def download(url, filename):
+def download(url):
+    response = requests.get(url, stream=True)
+    filename = parse_requests_response(response).filename_unsafe
+
+    default_filename = 'file'
+    if filename is None:
```
Comment (marco-c) [inline get_dependencies.py]: As I said, if we don't find the filename in the Content-Disposition header or as the basename of the URL, we want to throw an exception.

File: get_dependencies.py [hunk 12]
```diff
@@ -3,11 +3,21 @@
...
+import posixpath
 
 
-def download(url, filename):
+def download(url):
+    response = requests.get(url, stream=True)
+    filename = parse_requests_response(response).filename_unsafe
+    try:
+        filename = parse_requests_response(response).filename_unsafe
+    except:
```
Comment (amit2rockon) [inline get_dependencies.py]: what will be the exception?

File: requirements.txt [hunk 13]
```diff
@@ -4,3 +4,4 @@ keras==2.1.1
 tensorflow==1.3.0
 h5py==2.7.1
 requests==2.18.4
+
```
Comment (marco-c) [inline requirements.txt]: Please remove this empty line.

File: get_dependencies.py [hunk 14]
```diff
@@ -22,32 +32,42 @@ def download(url, filename):
...
-download('https://www.dropbox.com/s/nkf7a6jq13gmlnu/data.zip?dl=1', 'data.zip')
+download('https://www.dropbox.com/s/nkf7a6jq13gmlnu/data.zip?dl=1')
 
-print('[*] Extracting data.zip...')
 with ZipFile('data.zip', 'r') as z:
+    print('[*] Extracting data.zip...')
     z.extractall()
-
 os.remove('data.zip')
+
```
Comment (marco-c) [inline get_dependencies.py]: Please remove this empty line.

File: get_dependencies.py [hunk 15]
```diff
@@ -22,32 +32,42 @@ def download(url, filename):
...
 
 print('[*] Downloading data.zip...')
-download('https://www.dropbox.com/s/nkf7a6jq13gmlnu/data.zip?dl=1', 'data.zip')
+download('https://www.dropbox.com/s/nkf7a6jq13gmlnu/data.zip?dl=1')
 
-print('[*] Extracting data.zip...')
 with ZipFile('data.zip', 'r') as z:
+    print('[*] Extracting data.zip...')
     z.extractall()
-
```
Comment (marco-c) [inline get_dependencies.py]: Please don't remove this line.

File: requirements.txt (hunk 13 above)
Comment (amit2rockon) [reply requirements.txt]: okay

File: get_dependencies.py [hunk 16]
```diff
@@ -22,32 +32,42 @@ def download(url, filename):
...
     f.extractall('.')
 os.remove(name)
 
 print('[*] Downloading data.zip...')
-download('https://www.dropbox.com/s/nkf7a6jq13gmlnu/data.zip?dl=1', 'data.zip')
+download('https://www.dropbox.com/s/nkf7a6jq13gmlnu/data.zip?dl=1')
 
-print('[*] Extracting data.zip...')
 with ZipFile('data.zip', 'r') as z:
+    print('[*] Extracting data.zip...')
```
Comment (marco-c) [inline get_dependencies.py]: Why did you move the print statement in here?

File: get_dependencies.py [hunk 17]
```diff
@@ -22,32 +32,42 @@ def download(url, filename):
...
     url = 'https://www.dropbox.com/s/k4yifantsypy9xv/mac.tar.xz?dl=1'
-    name = 'mac.tar.xz'
 elif sys.platform.startswith('win32'):
     url = 'https://www.dropbox.com/s/xskj9rpn2fjkra8/win32.tar.xz?dl=1'
-    name = 'win32.tar.xz'
 
 print('[*] Downloading support files...')
-download(url, name)
 
+name = download(url)
```
Comment (marco-c) [inline get_dependencies.py]: Move this at the same line as it was before

File: get_dependencies.py [hunk 18]
```diff
@@ -3,11 +3,21 @@
...
 from zipfile import ZipFile
 import requests
+from rfc6266 import parse_requests_response
+import posixpath
 
 
-def download(url, filename):
+def download(url):
+    response = requests.get(url, stream=True)
+    filename = parse_requests_response(response).filename_unsafe
```
Comment (marco-c) [inline get_dependencies.py]: You're doing this twice. Do it only once without the try-except (if there is an exception, we want it to propagate to the caller).

File: get_dependencies.py [hunk 19]
```diff
@@ -22,32 +32,42 @@ def download(url, filename):
                 sys.stdout.write('\r[{}{}]'.format('█' * done, '.' * (50 - done)))
                 sys.stdout.flush()
     sys.stdout.write('\n')
+    return filename
+
+
+def sanitize(filename):
+    default_filename = 'file'
```
Comment (marco-c) [inline get_dependencies.py]: As I said before, we don't want a default filename. If there's no filename found, we want to throw an exception (`raise Exception('No filename could be found for this URL')`.

File: get_dependencies.py (hunk 18 above)
Comment (amit2rockon) [reply get_dependencies.py]: my code is like
    response = requests.get(url, stream=True)
    try:
        filename = parse_requests_response(response).filename_unsafe
    except Exception:
        print("Couldn\'t get file name for this URL")
         return 
what should I change here?

**Comment by amit2rockon** [comment]
@marco-c  any change required?

File: requirements.txt [hunk 20]
```diff
@@ -3,4 +3,4 @@ selenium==3.4.3
 keras==2.1.1
 tensorflow==1.3.0
 h5py==2.7.1
-requests==2.18.4
+requests==2.18.4
```
Comment (marco-c) [inline requirements.txt]: Can you remove the changes from this file?

File: get_dependencies.py [hunk 21]
```diff
@@ -3,11 +3,21 @@
...
 
 
-def download(url, filename):
+def download(url):
+    response = requests.get(url, stream=True)
+    filename = parse_requests_response(response).filename_unsafe
+
+    if filename is None:
+        raise Exception('No filename could be found for this URL')
+        return
```
Comment (marco-c) [inline get_dependencies.py]: No need to return after raising an exception

File: get_dependencies.py (hunk 18 above)
Comment (marco-c) [reply get_dependencies.py]: You've done it 👍

File: requirements.txt (hunk 20 above)
Comment (amit2rockon) [reply requirements.txt]: what is changed in this line?

File: get_dependencies.py (hunk 21 above)
Comment (amit2rockon) [reply get_dependencies.py]: okay

**Comment by amit2rockon** [comment]
@marco-c  changes has been made accordingly.

**Comment by marco-c** [comment]
Thanks!
